src/core/components/libp2p.js

```diff
@@ -13,15 +13,20 @@ module.exports = function libp2p (self, config) {
   const options = self._options || {}
   config = config || {}
 
-  // Always create libp2p via a bundle function
-  const createBundle = typeof options.libp2p === 'function'
-    ? options.libp2p
-    : defaultBundle
-
   const { datastore } = self._repo
   const peerInfo = self._peerInfo
   const peerBook = self._peerInfoBook
-  const libp2p = createBundle({ options, config, datastore, peerInfo, peerBook })
+
+  const libp2pOptions = getLibp2pOptions({ options, config, datastore, peerInfo, peerBook })
+  let libp2p
+
+  if (typeof options.libp2p === 'function') {
+    libp2p = options.libp2p({ libp2pOptions, options, config, datastore, peerInfo, peerBook })
+  } else {
+    // Required inline to reduce startup time
+    const Libp2p = require('libp2p')
+    libp2p = new Libp2p(mergeOptions(libp2pOptions, get(options, 'libp2p', {})))
+  }
```

Comment:
So, options is getting passed into `getLibp2pOptions`, but then it is still passed into libp2p creation. Should `getLibp2pOptions` just handle merging all of that so options doesn't need to be passed again?

> options.libp2p({ libp2pOptions, options, config, datastore, peerInfo, peerBook })

Along those lines, why is `options` being passed here? It feels like the result of `getLibp2pOptions` should just be able to get passed directly to both `new Libp2p` and `options.libp2p`. No other params should be needed right?

Reply:
> Should getLibp2pOptions just handle merging all of that so options doesn't need to be passed again?

There may be some other IPFS option you want to use in your custom bundle function. It's "passed again" because that's what currently happens, and I'm not trying to avoid a breaking change.

> > options.libp2p({ libp2pOptions, options, config, datastore, peerInfo, peerBook })
>
> Along those lines, why is `options` being passed here? It feels like the result of `getLibp2pOptions` should just be able to get passed directly to both `new Libp2p` and `options.libp2p`. No other params should be needed right?

`options` is IPFS options, passed to the IPFS constructor. It's passed here because that's currently what happens when you provide a custom bundle function. If we removed it, we'd have a breaking change.

`getLibp2pOptions` is passed IPFS `options` and returns _libp2p options_ (I would call them defaults, but they're not really because they're based off the IPFS `options` and the IPFS config from the repo).

Reply:
Okay, so `options` and `config` are for IPFS? If so, would it be easier to understand if in the scope of the custom bundler `options` were the libp2p options and the IPFS ones were prefixed? If those are going to be exposed I think it might be helpful to make that distinction.

Reply:
nvm, that would be a breaking change. I think that would be clearer, but since the goal is to avoid the breaking change I think this is fine.

Reply:
> I think that would be clearer

FWIW I agree!